src/ControlSystem/Averager.cpp

```diff
@@ -0,0 +1,162 @@
+}
+
+template <size_t DerivOrder>
+void Averager<DerivOrder>::update(const double time, const DataVector& raw_q,
+                                  const DataVector& timescales) noexcept {
+  if (not raw_qs_.empty()) {
+    // ensure that the number of components passed to update is the
+    // same as the number of components in the raw_q history
+    if (raw_q.size() != raw_qs_[0].size()) {
+      ERROR("The number of components in the raw_q provided ("
+            << raw_q.size()
+            << ") does not match the size of previously supplied raw_q ("
+            << raw_qs_[0].size() << ").");
+    }
+  } else {
+    // This is the first call to update: initialize averaged values, weights and
+    // effective time (with proper number of components)
+    averaged_values_ =
+        make_array<DerivOrder + 1>(DataVector(raw_q.size(), 0.0));
+    weight_k_ = averaged_values_[0];
+    tau_k_ = averaged_values_[0];
+  }
+
+  // update deques
+  times_.emplace_front(time);
+  raw_qs_.emplace_front(raw_q);
+  if (times_.size() > DerivOrder + 1) {
+    // get rid of old data once we have a sufficient number of points
+    times_.pop_back();
+    raw_qs_.pop_back();
+  }
+
+  if (times_.size() > DerivOrder) {
+    // we now have enough data points to begin averaging
+    const DataVector tau_avg = timescales * avg_tscale_fac_;
+    const DataVector tau_m{tau_avg.size(), times_[0] - times_[1]};
+
+    // update the weights and effective time
+    const DataVector old_weight{std::move(weight_k_)};
+    weight_k_ = (tau_m + old_weight) * tau_avg / (tau_m + tau_avg);
+    tau_k_ = (time * tau_m + tau_k_ * old_weight) * tau_avg /
+             (weight_k_ * (tau_m + tau_avg));
+
+    std::array<DataVector, DerivOrder + 1> raw_derivs = get_derivs();
+    for (size_t i = 0; i < averaged_values_.size(); i++) {
+      gsl::at(averaged_values_, i) =
+          (gsl::at(raw_derivs, i) * tau_m +
+           gsl::at(averaged_values_, i) * old_weight) *
+          tau_avg / (weight_k_ * (tau_m + tau_avg));
+    }
+    // replace averaged function value with raw value if
+    // not using `average_0th_deriv_of_q`
+    if (not average_0th_deriv_of_q_) {
+      averaged_values_[0] = raw_qs_[0];
+    }
+  }
+}
+
+template <size_t DerivOrder>
+double Averager<DerivOrder>::last_time_updated() const noexcept {
+  if (times_.empty()) {
+    ERROR("The time history has not been updated yet.");
+  }
+  return times_[0];
+}
+
+template <size_t DerivOrder>
+DataVector Averager<DerivOrder>::average_time(const double time) const
+    noexcept {
+  if (not have_valid_data(time)) {
+    ERROR(
+        "Cannot return averaged values because the averager does not have "
+        "up-to-date information.");
+  }
+  return tau_k_;
+}
+
+template <size_t DerivOrder>
+std::array<DataVector, DerivOrder + 1> Averager<DerivOrder>::get_derivs() const
+    noexcept {
+  static_assert(
+      DerivOrder == 2,
+      "Finite differencing currently only implemented for DerivOrder = 2");
```

Comment:
I'm still not happy with having a template parameter that must always be specified as 2, and would prefer that it were removed.  If we are going to keep it, it should be documented very clearly in the class documentation.